Sources/GRPC/ConnectionPool/ConnectionPool.swift

```diff
@@ -0,0 +1,796 @@
+
+    return (multiplexer, eventLoop)
+  }
+
+  /// Request a multiplexer for a single use when one becomes available in the future.
+  ///
+  /// - Parameters:
+  ///   - eventLoop: The `EventLoop` to be notified about the multiplexer on. The multiplexer which
+  ///       succeeds the returned `EventLoopFuture` may run on a different `EventLoop`.
+  ///   - deadline: The deadline by which the returned `EventLoopFuture` will be failed if no
+  ///       multiplexer has become available. If this is not provided then the deadline will be
+  ///       `maximumConnectionWaitTime` from now.
+  /// - Returns: A future `HTTP2StreamMultiplexer`.
+  internal func waitForMultiplexer(
+    eventLoop: EventLoop,
+    until deadline: NIODeadline? = nil
+  ) -> EventLoopFuture<HTTP2StreamMultiplexer> {
+    let promise = eventLoop.makePromise(of: HTTP2StreamMultiplexer.self)
+    let actualDeadline = deadline ?? .now() + self.maximumConnectionWaitTime
+
+    self.lock.withLockVoid {
+      self._waitForMultiplexer(promise: promise, until: actualDeadline)
+    }
+
+    return promise.futureResult
+  }
+
+  /// Lock-held implementation of `waitForMultiplexer(eventLoop:deadline:)`.
+  private func _waitForMultiplexer(
+    promise: EventLoopPromise<HTTP2StreamMultiplexer>,
+    until deadline: NIODeadline
+  ) {
+    guard case .active = self.state else {
+      // Shutdown or shutting down, not much we can do.
+      promise.fail(ConnectionPoolError.shutdown)
+      return
+    }
+
+    guard self.connectionWaiters.count < self.maximumConnectionWaiters else {
+      // The connection waiter queue is full: avoid overwhelming the pool.
+      promise.fail(ConnectionPoolError.tooManyWaiters)
+      return
+    }
+
+    // Enqueue a waiter.
+    let waiter = self.makeWaiter(promise: promise, deadline: deadline)
+    self.connectionWaiters.append(waiter)
+
+    // We're waiting for a multiplexer, chances are we don't have any capacity on existing ones,
+    // request a connection.
+    self._requestConnection()
+
+    self.logger.trace("Enqueued connection waiter", metadata: [
+      "pool.waiters.count": "\(self.connectionWaiters.count)",
+    ])
+  }
+
+  /// Make a connection waiter, fulfilling the given promise with a multiplexer if one becomes
+  /// available before the `deadline`.
+  ///
+  /// - Parameters:
+  ///   - promise: The promise to complete when a multiplexer becomes available.
+  ///   - deadline: The deadline to wait until before giving up on waiting for a multiplexer.
+  /// - Returns: A `Waiter`.
+  private func makeWaiter(
+    promise: EventLoopPromise<HTTP2StreamMultiplexer>,
+    deadline: NIODeadline
+  ) -> Waiter {
+    var waiter = Waiter(multiplexerPromise: promise)
+    let waiterID = waiter.id
+
+    // Schedule a timeout (executed on the event loop) to fail the promise.
+    waiter.scheduleTimeout(at: deadline, on: promise.futureResult.eventLoop) {
+      self.timeoutWaiter(withID: waiterID)
+    }
+
+    return waiter
+  }
```

Comment:
Good point!

Reply:
Actually, I'm not sure this is right since the waiter stores the `Scheduled` task; if we schedule the deadline after releasing the lock then the waiter stored in `connectionWaiters` won't have a scheduled task to cancel when it is succeeded/failed. Am I missing something?

Reply:
No, you aren’t. That means actually doing the schedule has to happen during the critical section. That’s somewhat painful, but acceptable for now I think.